src/Conversion/ONNXToTOSA/Tensor/Concat.cpp

```diff
@@ -0,0 +1,71 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//===---------------- Concat.cpp - Concat Op --------------------===//
+//
+// Copyright (c) 2023 Advanced Micro Devices, Inc.
+//
+// =============================================================================
+//
+// This file lowers ONNX ConcatOp to TOSA dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/Tosa/IR/TosaOps.h"
+#include "src/Conversion/ONNXToTOSA/ONNXToTOSACommon.hpp"
+#include "src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.hpp"
+#include "src/Dialect/ONNX/ONNXOps.hpp"
+
+using namespace mlir;
+
+namespace onnx_mlir {
+
+namespace {
+
+class ONNXConcatLoweringToTOSA : public OpConversionPattern<ONNXConcatOp> {
+public:
+  using OpConversionPattern::OpConversionPattern;
+  using OpAdaptor = typename ONNXConcatOp::Adaptor;
+  LogicalResult matchAndRewrite(ONNXConcatOp op, OpAdaptor adaptor,
+      ConversionPatternRewriter &rewriter) const override {
+
+    ValueRange inputs = adaptor.inputs();
+    int64_t axis = adaptor.axis();
+    auto resultType = op.getResult().getType();
+
+    for (const auto &input : inputs) {
+      if (!onnx_mlir::isRankedShapedType(input.getType()))
+        return rewriter.notifyMatchFailure(
+            op, "inputs are not ranked shaped tensors");
+    }
+    int64_t inputRank = onnx_mlir::getRank(inputs[0].getType());
+
+    // onnx allows values beetween [-r, r-1] where r is the rank.
+    if (axis < 0)
+      axis += inputRank;
+
+    // Check if axis is in correct range.
+    if (axis < 0 || axis >= inputRank)
+      return rewriter.notifyMatchFailure(
+          op, "axis attribute not in correct range");
+
+    Type newConcatOutputType =
```

Comment:
I don't know what this type is used for. I expected the result of the tosa op to be the same as the result of the onnx op, but assume I'm missing something.

Reply:
It should be. This is some nice behavior I encountered. By letting shape inference fill out the shape of the concat (or the other ops where I used this pattern), we ensure that the shape calculations and related attributes of the Tosa Op are correct. If the shape inference finds another shape, the pattern will fail with a useful error message.

Reply:
This is for example the way how I found the `ceil` error in maxpool